### PR TITLE
Fix a number of issues with the gpio.pulse family of functions

### LINKED ISSUE #2254: Issues with gpio.pulse
### Expected behavior
Handling of adjusts should work even when value is large
### Actual behavior
It gets confused if the new delay exceeds min or max

### NodeMCU version
dev

Also, increase the maximum value of delay above one second.

### PR DESCRIPTION
**Pull request by pjsg**
Fixes #2254.

- [x] This PR is for the `dev` branch rather than for `master`.
- [x] This PR is compliant with the [other contributing guidelines](https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md) as well (if not, please describe why).
- [x] I have thoroughly tested my contribution.
- [x] The code changes are reflected in the documentation at `docs/en/*`.

I'm now trying to make progress on my clock project and discovering some bugs in the existing code, and missing functionality. This PR adds the missing features and resolves the bugs.

### CODE REVIEW

File: app/modules/gpio_pulse.c [hunk 1]
```diff
@@ -137,6 +137,69 @@ static int gpio_pulse_delete(lua_State *L) {
...
+  if (lua_type(L, -1) != LUA_TTABLE) {
+    luaL_error(L, "All entries must be tables");
+  }
+
+  lua_pushnil(L);             // key position
+  while (lua_next(L, -2)) {
+    // stack now contains: -1 => value; -2 => key; -3 => table
+    if (lua_type(L, -2) == LUA_TNUMBER) {
+      int pin = lua_tonumber(L, -2);
+      int value = lua_tonumber(L, -1);
```
Comment (devsaurus) [inline app/modules/gpio_pulse.c]: What if value at -1 isn't LUA_TNUMBER or a convertible string?

File: app/modules/gpio_pulse.c [hunk 2]
```diff
@@ -137,6 +137,69 @@ static int gpio_pulse_delete(lua_State *L) {
...
+        entry->gpio_set |= BIT(pin_num[pin]);
+      } else {
+        entry->gpio_clr |= BIT(pin_num[pin]);
+      }
+    } else {
+      const char *str = lua_tostring(L, -2);
+
+      if (strcmp(str, "delay") == 0) {
+        entry->delay = lua_tonumber(L, -1);
+        if (entry->delay < 0 || entry->delay > DELAY_LIMIT) {
```
Comment (devsaurus) [inline app/modules/gpio_pulse.c]: I expect that this won't throw an error if value at index `-1` is not convertible to a number. [`lua_tonumber()`](http://pgl.yoyo.org/luai/i/lua_tonumber) returns 0 in this case and the range check is happy.

`luaL_checknumber()` or an explicit test for `lua_isnumber()` if delay == 0 would detect nonsense.

Applies to further instances of `lua_tonumber()` below.

**Comment by pjsg** [comment]
The changes aren't as big as they look as I moved the guts of one function into its own function so that I could use it elsewhere.

I did fix all the unchecked calls so that there is more error detection (and it actually found an error in my current application!)

**Comment by pjsg** [comment]
This also includes a fix to the issue in https://stackoverflow.com/q/48790455/131929 -- I changed gpio.mode(, gpio.OUTPUT) to actually enable the driver.

**Comment by pjsg** [comment]
I also addressed the comments in #2265

**Pull request merged**